vnext/Shared/OInstance.cpp

```diff
@@ -458,13 +479,35 @@ void InstanceImpl::loadBundleInternal(std::string &&jsBundleRelativePath, bool s
           m_devSettings->platformName,
           /*dev*/ true,
           /*hot*/ false,
-          m_devSettings->inlineSourceMap);
-
-      // Remote debug executor loads script from a Uri, rather than taking the actual bundle string
-      m_innerInstance->loadScriptFromString(
-          std::make_unique<const JSBigStdString>(m_devSettings->useWebDebugger ? bundleUrl : jsBundleString),
-          bundleUrl,
-          synchronously);
+          m_devSettings->inlineSourceMap,
+          hermesBytecodeVersion);
+
+      // This code is based on the HBC Bundle integration on Android
+      // Ref:
+      // https://github.com/facebook/react-native/blob/24d91268b64c7abbd4b26547ffcc663dc90ec5e7/ReactAndroid/src/main/jni/react/jni/CatalystInstanceImpl.cpp#L231
+      if (isHBCBundle(jsBundleString.c_str())) {
+        auto script = std::make_unique<JSBigStdString>(jsBundleString, false);
+        const char *buffer = script->c_str();
+        uint32_t bufferLength = (uint32_t)script->size();
+        uint32_t offset = 8;
+        while (offset < bufferLength) {
```

Comment:
is this condition correct, if we are going to access `offset + 4`?

Reply:
As explained in the preceding comment, each module chunk starts with 4 bytes which encodes module length .. This condition checks whether the next module chunk starts within the buffer ... if (offset + 4) overflows, then moduleLength will be computed as '0' and we end up evaluating an empty script and then break ... But the JSBigStdString creation can result in AV if constructing std::string with same start and end address (degenerate case?) tries to access the memory .. Hence it seems safer to exit early if module length is zero .. This condition shouldn't arise for a well formatted HBCB though.

Reply:
right. Who provides the HBCB? is this something that will ever be used in production, and is this something an attacker can try to provide? I'm not concerned about the well-formatted well-intentioned hbcb, but about someone actively trying to buffer overrun to exploit the app through a buffer overrun

Reply:
HBCB is never supposed to be used in production .. It essentially compiles each modules (prior to bundling) independently and append them together .. it enables recompiling and serving only changed modules through dev server which is much faster than bundling and compiling the whole bundle .. 

We handle the HBCB only when the FastRefresh is enabled and the bundle is coming downloaded from the devserver. If an attacker can change enable the FastRefresh flag in devsettings and run a bundle from devserver in production, they could exploit HBCB load path.